### PR TITLE
Canaries via Kafka

### PR DESCRIPTION
**Pull request by SystemFw**
This PR switches the canaries listening mechanism from polling to Kafka.
It also addresses a large swath of technical debt, both on the code and the design of the project.
You will see a few TODO items that needs addressing. In particular, at the moment there is no logging. While I'm not a fan of logging useless things everywhere, we definitely need more than zero, so if you could help point out the main things we want to log, that would be helpful. 
Same goes for error handling, it needs to be a bit more tight, although at the moment the streams should restart on error.

### CODE REVIEW

File: src/main/scala/com/ovoenergy/comms/monitor/canary/logic.scala [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+import com.ovoenergy.comms.model.{Feedback, FeedbackOptions}, FeedbackOptions.Failed
+
+object logic {
+
+  def apply[F[_]: Concurrent: Timer](
+    config: Config,
+    kafka: Kafka
+  ): Stream[F, Unit] = {
+    val awsRegion = config.getString("aws.region")
```
Comment (filosganga) [inline src/main/scala/com/ovoenergy/comms/monitor/canary/logic.scala]: Do we really need to specify the region to create the AmazonSNS? I believe it should be resolved by the `DefaultAWSRegionSupplierChain` or something similar.

File: src/main/scala/com/ovoenergy/comms/monitor/pipeline/metrics/kafka/TopicsConsumerStream.scala [hunk 2]
```diff
@@ -38,24 +31,22 @@ object TopicsConsumerStream extends LoggingWithMDC {
...
-    val composedEmailV4Topic: Topic[ComposedEmailV4] = aivenCluster.composedEmail.v4
-    val composedSmsV4Topic: Topic[ComposedSMSV4] = aivenCluster.composedSms.v4
-    val composedPrintV2Topic: Topic[ComposedPrintV2] = aivenCluster.composedPrint.v2
-    val progressedEmailV3Topic: Topic[EmailProgressedV3] = aivenCluster.progressedEmail.v3
-    val progressedSmsV3Topic: Topic[SMSProgressedV3] = aivenCluster.progressedSMS.v3
-    val issuedForDeliveryV3Topic: Topic[IssuedForDeliveryV3] = aivenCluster.issuedForDelivery.v3
-    val failedV3Topic: Topic[FailedV3] = aivenCluster.failed.v3
+  def apply[F[_]: Concurrent](kafka: Kafka)(f: Record[MetricsMessage] => F[Unit]) = {
+
+    // TODO this is full of side-effects (`.deserializer`) and duplication
```
Comment (filosganga) [inline src/main/scala/com/ovoenergy/comms/monitor/pipeline/metrics/kafka/TopicsConsumerStream.scala]: Yes, it is! 

we should:
 - Decouple Kafka configuration from typesafe config
 - Get rid of comms-kafka-serialisation
 - Break the concept of one-topic-one-consumer that is baked in the Topic.

File: src/main/scala/com/ovoenergy/comms/monitor/canary/model.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      }
+    }
+  }
+
+  case class TraceToken(value: String)
+  object TraceToken {
+    implicit def eqTokens: Eq[TraceToken] = Eq.by(_.value)
+  }
+
+  // TODO add timestamp and timeout expected/ttl?
```
Comment (filosganga) [inline src/main/scala/com/ovoenergy/comms/monitor/canary/model.scala]: Good idea!!!

File: src/main/scala/com/ovoenergy/comms/monitor/canary/logic.scala (hunk 1 above)
Comment (SystemFw) [reply src/main/scala/com/ovoenergy/comms/monitor/canary/logic.scala]: I have no idea, but if we could avoid that's certainly a plus

**Comment by SystemFw** [comment]
With the feedback topic story done, this is ready for final review/merge

File: src/main/scala/com/ovoenergy/comms/monitor/canary/logic.scala [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    kafka: Kafka,
+    logger: Logger[F]
+  ): Stream[F, Unit] = {
+    val awsRegion = config.getString("aws.region")
+    val snsTopicTriggeredApiV2 = config.getString("aws.canary.sns_topics.api.v2")
+    val snsTopicTriggeredApiV3 = config.getString("aws.canary.sns_topics.api.v3")
+    val pagerDutyApiKey = config.getString("canary.alert.pagerduty.apiKey")
+
+    def log(id: TraceToken): Logger[F] =
+      logger.withModifiedString { msg =>
```
[the remark this answers is not in the document]
Comment (SystemFw) [reply src/main/scala/com/ovoenergy/comms/monitor/canary/logic.scala]: We need to figure out the exact format to make this parseable by logz.io 
Definitely a point for the logging spike

**Comment by ovo-comms-circleci** [comment]
This PR was successfully deployed to production by [CircleCI build 398](https://circleci.com/gh/ovotech/comms-monitor-service/398) at 2018-09-03 15:15 BST.

**Pull request merged**